confidential-datahub/kms/src/api.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//! # Api definitions for KMS/Vault drivers
+//!
+//! To establish a connection between the client and the KMS/Vault server, two
+//! kinds of parameters are required:
+//! - Public parameters: like Region Id of the KMS/Vault, Instance Id of the KMS,
+//! etc. They are not confidential and can be seen by anyone. [`ProviderSetting`]
+//! is a json object. It is to include all the public parameters. The
+//! hashmap style makes it flexible for different kinds of KMSes/Vaults. There
+//! should be a function that takes a [`ProviderSetting`] as a input parameter
+//! and returns a client instance.
+//! - Private parameters: like the credential to access (e.g. access key).
+//! These parameters should be captured inside the logic of `new()` rather
+//! than the input parameter. it is strongly recommended that private parameters
+//! be read from the encrypted filesystem, e.g. `/run/*` which is in TEE's
+//! encrypted memory.
+//!
+//! ## APIs
+//! - `Decryptor`: KMS's decrypt API.
+//! - `Encryptor`: KMS's encrypt API.
+//! - `Getter`: Vault's get secret API.
+//! - `Setter`: Vault's set secret API.
+//!
+//! The rationality to distinguish these four different traits:
+//! - `Decryptor` and `Getter` are used in-pod, while `Encryptor` and `Setter`
+//! are used userside. They do not need to be implemented by a same object.
```

Comment:
Do we plan to implement a secret sealing tool somewhere for the client? Will it share any of this code?

Reply:
Yes. That tool will leverage `Encrypter` trait. We will put that in later. The secret sealing tool is some higher level codes which calls the `kms` crate.

For aliyun, it will also use the `AliyunKmsClient`.